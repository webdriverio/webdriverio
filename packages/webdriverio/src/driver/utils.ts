import cp from 'node:child_process'

import decamelize from 'decamelize'
import logger from '@wdio/logger'
import { getChromePath } from 'chrome-launcher'
import {
    install, canDownload, resolveBuildId, detectBrowserPlatform, Browser, ChromeReleaseChannel,
    computeExecutablePath, type InstallOptions
} from '@puppeteer/browsers'
import type { EdgedriverParameters } from 'edgedriver'
import type { Capabilities } from '@wdio/types'

const log = logger('webdriver')
const EXCLUDED_PARAMS = ['version', 'help']

export function parseParams(params: EdgedriverParameters) {
    return Object.entries(params)
        .filter(([key,]) => !EXCLUDED_PARAMS.includes(key))
        .map(([key, val]) => {
            if (typeof val === 'boolean' && !val) {
                return ''
            }
            const vals = Array.isArray(val) ? val : [val]
            return vals.map((v) => `--${decamelize(key, { separator: '-' })}${typeof v === 'boolean' ? '' : `=${v}`}`)
        })
        .flat()
        .filter(Boolean)
}

export function getLocalChromePath() {
    try {
        const chromePath = getChromePath()
        return chromePath
    } catch (err: unknown) {
        return
    }
}

export function getBuildIdByPath(chromePath?: string) {
    if (!chromePath) {
        return
    }
    const versionString = cp.execSync(`"${chromePath}" --version`).toString()
    return versionString.split(' ').pop()?.trim()
}

export async function setupChrome(caps: Capabilities.Capabilities, cacheDir: string) {
    const platform = detectBrowserPlatform()
    if (!platform) {
        throw new Error('The current platform is not supported.')
    }

    if (!caps.browserVersion) {
        const executablePath = getLocalChromePath()!
        const tag = getBuildIdByPath(executablePath)

        /**
         * verify that we have a valid Chrome browser installed
         */
        if (tag) {
            return {
                platform,
                executablePath,
                buildId: await resolveBuildId(Browser.CHROME, platform, tag)
            }
        }
    }

    /**
     * otherwise download provided Chrome browser version or "stable"
     */
    const tag = caps.browserVersion || ChromeReleaseChannel.STABLE
    const buildId = await resolveBuildId(Browser.CHROME, platform, tag)
    const installOptions: InstallOptions = {
        cacheDir,
        platform,
        buildId,
        browser: Browser.CHROME,
        downloadProgressCallback: () => {}
    }
    const isCombinationAvailable = await canDownload(installOptions)
    if (!isCombinationAvailable) {
        throw new Error(`Couldn't find a matching Chrome browser for tag "${buildId}" on platform "${platform}"`)
    }

    log.info(`Downloading Chrome v${buildId}`)
    await install(installOptions)
    const executablePath = computeExecutablePath(installOptions)
    return { executablePath, buildId, platform }
}
