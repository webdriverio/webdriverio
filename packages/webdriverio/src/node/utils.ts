import path from 'node:path'
import fs from 'node:fs/promises'

/**
 * check if directory exists
 */
export async function assertDirectoryExists(filepath: string) {
    const exist = await fs.access(path.dirname(filepath)).then(() => true, () => false)
    if (!exist) {
        throw new Error(`directory (${path.dirname(filepath)}) doesn't exist`)
    }
}

/**
 * Get absolute file path of a file
 * @param filepath given file path
 * @returns absolute file path
 */
export function getAbsoluteFilepath(filepath: string) {
    return filepath.startsWith('/') || filepath.startsWith('\\') || filepath.match(/^[a-zA-Z]:\\/)
        ? filepath
        : path.join(process.cwd(), filepath)
}
