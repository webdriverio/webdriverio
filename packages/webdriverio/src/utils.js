import fs from 'fs'
import path from 'path'
import cssValue from 'css-value'
import rgb2hex from 'rgb2hex'
import GraphemeSplitter from 'grapheme-splitter'

import { ELEMENT_KEY, W3C_SELECTOR_STRATEGIES, UNICODE_CHARACTERS } from './constants'

const DEFAULT_SELECTOR = 'css selector'
const DIRECT_SELECTOR_REGEXP = /^(id|css selector|xpath|link text|partial link text|name|tag name|class name|-android uiautomator|-ios uiautomation|accessibility id):(.+)/
const INVALID_SELECTOR_ERROR = new Error('selector needs to be typeof `string` or `function`')

export const findStrategy = function (value, isW3C, isMobile) {
    /**
     * set default selector
     */
    let using = DEFAULT_SELECTOR

    /**
     * check if user has specified locator strategy directly
     */
    const match = value.match(DIRECT_SELECTOR_REGEXP)
    if (match) {
        /**
         * ensure selector strategy is supported
         */
        if (!isMobile && isW3C && !W3C_SELECTOR_STRATEGIES.includes(match[1])) {
            throw new Error('InvalidSelectorStrategy') // ToDo: move error to wdio-error package
        }

        return {
            using: match[1],
            value: match[2]
        }
    }

    // use xPath strategy if value starts with //
    if (value.indexOf('/') === 0 || value.indexOf('(') === 0 ||
               value.indexOf('../') === 0 || value.indexOf('./') === 0 ||
               value.indexOf('*/') === 0) {
        using = 'xpath'

    // use link text strategy if value starts with =
    } else if (value.indexOf('=') === 0) {
        using = 'link text'
        value = value.slice(1)

    // use partial link text strategy if value starts with *=
    } else if (value.indexOf('*=') === 0) {
        using = 'partial link text'
        value = value.slice(2)

    // recursive element search using the UiAutomator library (Android only)
    } else if (value.indexOf('android=') === 0) {
        using = '-android uiautomator'
        value = value.slice(8)

    // recursive element search using the UIAutomation library (iOS-only)
    } else if (value.indexOf('ios=') === 0) {
        using = '-ios uiautomation'
        value = value.slice(4)

    // recursive element search using accessibility id
    } else if (value.indexOf('~') === 0) {
        using = 'accessibility id'
        value = value.slice(1)

    // class name mobile selector
    // for iOS = UIA...
    // for Android = android.widget
    } else if (value.slice(0, 3) === 'UIA' || value.slice(0, 15) === 'XCUIElementType' || value.slice(0, 14).toLowerCase() === 'android.widget') {
        using = 'class name'

    // use tag name strategy if value contains a tag
    // e.g. "<div>" or "<div />"
    } else if (value.search(/<[a-zA-Z-]+( \/)*>/g) >= 0) {
        using = 'tag name'
        value = value.replace(/<|>|\/|\s/g, '')

    // use name strategy if value queries elements with name attributes for JSONWP
    // e.g. "[name='myName']" or '[name="myName"]'
    } else if (!isW3C && value.search(/^\[name=("|')([a-zA-z0-9\-_. ]+)("|')]$/) >= 0) {
        using = 'name'
        value = value.match(/^\[name=("|')([a-zA-z0-9\-_. ]+)("|')]$/)[2]

    // allow to move up to the parent or select current element
    } else if (value === '..' || value === '.') {
        using = 'xpath'

    // any element with given class, id, or attribute and content
    // e.g. h1.header=Welcome or [data-name=table-row]=Item or #content*=Intro
    } else {
        const match = value.match(new RegExp([
            // HTML tag
            /^([a-z0-9|-]*)/,
            // optional . or # + class or id
            /(?:(\.|#)(-?[_a-zA-Z]+[_a-zA-Z0-9-]*))?/,
            // optional [attribute-name="attribute-value"]
            /(?:\[(-?[_a-zA-Z]+[_a-zA-Z0-9-]*)(?:=(?:"|')([a-zA-z0-9\-_. ]+)(?:"|'))?\])?/,
            // *=query or =query
            /(\*)?=(.+)$/,
        ].map(rx => rx.source).join('')))

        if (match) {
            const PREFIX_NAME = { '.': 'class', '#': 'id' }
            const conditions = []
            const [
                tag,
                prefix, name,
                attrName, attrValue,
                partial, query
            ] =  match.slice(1)

            if (prefix) {
                conditions.push(`contains(@${PREFIX_NAME[prefix]}, "${name}")`)
            }
            if (attrName) {
                conditions.push(
                    attrValue
                        ? `contains(@${attrName}, "${attrValue}")`
                        : `@${attrName}`
                )
            }
            if (partial) {
                conditions.push(`contains(., "${query}")`)
            } else {
                conditions.push(`normalize-space() = "${query}"`)
            }

            using = 'xpath'
            value = `.//${tag || '*'}[${conditions.join(' and ')}]`
        }
    }

    /**
     * ensure selector strategy is supported
     */
    if(!isMobile && isW3C && !W3C_SELECTOR_STRATEGIES.includes(using)){
        throw new Error('InvalidSelectorStrategy') // ToDo: move error to wdio-error package
    }

    return { using, value }
}

const applyScopePrototype = (prototype, scope) => {
    const dir = path.resolve(__dirname, 'commands', scope)
    const files = fs.readdirSync(dir)
    for (let filename of files) {
        const commandName = path.basename(filename, path.extname(filename))
        prototype[commandName] = { value: require(path.join(dir, commandName)).default }
    }
}

/**
 * enhances objects with element commands
 */
export const getPrototype = (scope) => {
    const prototype = {}

    /**
     * register action commands
     */
    applyScopePrototype(prototype, scope)
    return prototype
}

/**
 * get element id from WebDriver response
 * @param  {object} res         body object from response
 * @return {string|undefined}   element id or null if element couldn't be found
 */
export const getElementFromResponse = (res) => {
    /**
     * deprecated JSONWireProtocol response
     */
    if (res.ELEMENT) {
        return res.ELEMENT
    }

    /**
     * W3C WebDriver response
     */
    if (res[ELEMENT_KEY]) {
        return res[ELEMENT_KEY]
    }

    return null
}

/**
 * traverse up the scope chain until browser element was reached
 */
export function getBrowserObject (elem) {
    return elem.parent ? getBrowserObject(elem.parent) : elem
}

/**
 * transform whatever value is into an array of char strings
 */
export function transformToCharString (value) {
    const ret = []

    if (!Array.isArray(value)) {
        value = [value]
    }

    for (const val of value) {
        if (typeof val === 'string') {
            ret.push(...checkUnicode(val))
        } else if (typeof val === 'number') {
            const entry = `${val}`.split('')
            ret.push(...entry)
        } else if (val && typeof val === 'object') {
            try {
                ret.push(...JSON.stringify(val).split(''))
            } catch (e) { /* ignore */ }
        } else if (typeof val === 'boolean') {
            const entry = val ? 'true'.split('') : 'false'.split('')
            ret.push(...entry)
        }
    }

    return ret
}

function sanitizeCSS (value) {
    /* istanbul ignore next */
    if (!value) {
        return value
    }

    return value.trim().replace(/'/g, '').replace(/"/g, '').toLowerCase()
}

/**
 * parse css values to a better format
 * @param  {Object} cssPropertyValue result of WebDriver call
 * @param  {String} cssProperty      name of css property to parse
 * @return {Object}                  parsed css property
 */
export function parseCSS (cssPropertyValue, cssProperty) {
    if (!cssPropertyValue) {
        return null
    }

    let parsedValue = {
        property: cssProperty,
        value: cssPropertyValue.toLowerCase().trim()
    }

    if (parsedValue.value.indexOf('rgb') === 0) {
        /**
         * remove whitespaces in rgb values
         */
        parsedValue.value = parsedValue.value.replace(/\s/g, '')

        /**
         * parse color values
         */
        let color = parsedValue.value
        parsedValue.parsed = rgb2hex(parsedValue.value)
        parsedValue.parsed.type = 'color'
        parsedValue.parsed[/[rgba]+/g.exec(color)[0]] = color
    } else if (parsedValue.property === 'font-family') {
        let font = cssValue(cssPropertyValue)
        let string = parsedValue.value
        let value = cssPropertyValue.split(/,/).map(sanitizeCSS)

        parsedValue.value = sanitizeCSS(font[0].value || font[0].string)
        parsedValue.parsed = { value, type: 'font', string }
    } else {
        /**
         * parse other css properties
         */
        try {
            parsedValue.parsed = cssValue(cssPropertyValue)

            if (parsedValue.parsed.length === 1) {
                parsedValue.parsed = parsedValue.parsed[0]
            }

            if (parsedValue.parsed.type && parsedValue.parsed.type === 'number' && parsedValue.parsed.unit === '') {
                parsedValue.value = parsedValue.parsed.value
            }
        } catch (e) {
            // TODO improve css-parse lib to handle properties like
            // `-webkit-animation-timing-function :  cubic-bezier(0.25, 0.1, 0.25, 1)
        }
    }

    return parsedValue
}

/**
 * check for unicode character or split string into literals
 * @param  {String} value  text
 * @return {Array}         set of characters or unicode symbols
 */
export function checkUnicode (value) {
    return UNICODE_CHARACTERS.hasOwnProperty(value) ? [UNICODE_CHARACTERS[value]] : new GraphemeSplitter().splitGraphemes(value)
}

function fetchElementByJSFunction (selector, scope) {
    if (!scope.elementId) {
        return scope.execute(selector)
    }

    const script = ((elem) => (selector).call(elem)).toString().replace('selector', `(${selector.toString()})`)
    return getBrowserObject(scope).execute(`return (${script}).apply(null, arguments)`, scope)
}

/**
 * logic to find an element
 */
export async function findElement(selector) {
    /**
     * fetch element using regular protocol command
     */
    if (typeof selector === 'string') {
        const { using, value } = findStrategy(selector, this.isW3C, this.isMobile)
        return this.elementId
            ? this.findElementFromElement(this.elementId, using, value)
            : this.findElement(using, value)
    }

    /**
     * fetch element with JS function
     */
    if (typeof selector === 'function') {
        const notFoundError = new Error(`Function selector "${selector.toString()}" did not return an HTMLElement`)
        let elem = await fetchElementByJSFunction(selector, this)
        elem = Array.isArray(elem) ? elem[0] : elem
        return getElementFromResponse(elem) ? elem : notFoundError
    }

    throw INVALID_SELECTOR_ERROR
}

/**
 * logic to find a elements
 */
export async function findElements(selector) {
    /**
     * fetch element using regular protocol command
     */
    if (typeof selector === 'string') {
        const { using, value } = findStrategy(selector, this.isW3C, this.isMobile)
        return this.elementId
            ? this.findElementsFromElement(this.elementId, using, value)
            : this.findElements(using, value)
    }

    /**
     * fetch element with JS function
     */
    if (typeof selector === 'function') {
        let elems = await fetchElementByJSFunction(selector, this)
        elems = Array.isArray(elems) ? elems : [elems]
        return elems.filter((elem) => elem && getElementFromResponse(elem))
    }

    throw INVALID_SELECTOR_ERROR
}
