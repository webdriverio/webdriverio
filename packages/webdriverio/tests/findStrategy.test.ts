import { findStrategy } from '../src/utils/findStrategy'
import fs from 'fs'

jest.mock('fs')

describe('selector strategies helper', () => {
    it('should find an element using "css selector" method', () => {
        const element = findStrategy('.red')
        expect(element.using).toBe('css selector')
        expect(element.value).toBe('.red')
    })

    it('should find an element using "id" method', () => {
        const element = findStrategy('#purplebox')
        expect(element.using).toBe('css selector')
        expect(element.value).toBe('#purplebox')
    })

    it('should find an element using "name" method through jsonwp', () => {
        const element = findStrategy('[name="searchinput"]', false)
        expect(element.using).toBe('name')
        expect(element.value).toBe('searchinput')
    })

    it('should find an element using "name" method through WC3', () => {
        const element = findStrategy('[name="searchinput"]', true)
        expect(element.using).toBe('css selector')
        expect(element.value).toBe('[name="searchinput"]')
    })

    it('should find an element using "name" method with a . in the name', () => {
        const element = findStrategy('[name="search.input"]')
        expect(element.using).toBe('name')
        expect(element.value).toBe('search.input')
    })

    it('should find an element using "name" method with an attribute', () => {
        const element = findStrategy('[name="searchinput[@isDisplayed=\'true\']"]')
        expect(element.using).toBe('name')
        expect(element.value).toBe('searchinput[@isDisplayed=\'true\']')
    })

    it('should find an element using "name" method by "name" strategy if isMobile is used even when w3c is used', () => {
        const element = findStrategy('[name="searchinput"]', true, true)
        expect(element.using).toBe('name')
        expect(element.value).toBe('searchinput')
    })

    it('should find an element using "link text" method', () => {
        const element = findStrategy('=GitHub Repo')
        expect(element.using).toBe('link text')
        expect(element.value).toBe('GitHub Repo')
    })

    it('should find an element using "partial link text" method', () => {
        const element = findStrategy('*=new')
        expect(element.using).toBe('partial link text')
        expect(element.value).toBe('new')
    })

    it('should find an element using "id" method', () => {
        const element = findStrategy('id=purplebox')
        expect(element.using).toBe('id')
        expect(element.value).toBe('purplebox')
    })

    it('should find an element using "tag name" method and tag format <XXX />', () => {
        const element = findStrategy('<textarea />')
        expect(element.using).toBe('tag name')
        expect(element.value).toBe('textarea')
    })

    it('should find an element using "tag name" method and tag format <XXX>', () => {
        const element = findStrategy('<textarea>')
        expect(element.using).toBe('tag name')
        expect(element.value).toBe('textarea')
        const heading = findStrategy('<h2>')
        expect(heading.using).toBe('tag name')
        expect(heading.value).toBe('h2')
    })

    it('should find an element using "xpath" method', () => {
        const element = findStrategy('//html/body/section/div[6]/div/span')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('//html/body/section/div[6]/div/span')
    })

    it('should find an element using "xpath" method for ParenthesizedExpressions', () => {
        const element = findStrategy('(//div)[7]/span')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('(//div)[7]/span')
    })

    it('should find an element by tag name + content', () => {
        const element = findStrategy('div=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//div[normalize-space() = "some random text with "§$%&/()div=or others"]')
    })

    it('should find an element by tag name + id + similar content', () => {
        const element = findStrategy('h1=Christian')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//h1[normalize-space() = "Christian"]')
    })

    it('should find an element by tag name + similar content', () => {
        const element = findStrategy('div*=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//div[contains(., "some random text with "§$%&/()div=or others")]')
    })

    it('should find an element by tag name + class + content', () => {
        const element = findStrategy('div.some-class=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//div[contains(@class, "some-class") and normalize-space() = "some random text with "§$%&/()div=or others"]')
    })

    it('should find an element class + content', () => {
        const element = findStrategy('.some-class=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//*[contains(@class, "some-class") and normalize-space() = "some random text with "§$%&/()div=or others"]')
    })

    it('should find an element by tag name + class + similar content', () => {
        const element = findStrategy('div.some-class*=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//div[contains(@class, "some-class") and contains(., "some random text with "§$%&/()div=or others")]')
    })

    it('should find an element by class + similar content', () => {
        const element = findStrategy('.some-class*=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//*[contains(@class, "some-class") and contains(., "some random text with "§$%&/()div=or others")]')
    })

    it('should find an element by tag name + id + content', () => {
        const element = findStrategy('div#some-class=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//div[contains(@id, "some-class") and normalize-space() = "some random text with "§$%&/()div=or others"]')
    })

    it('should find an element by id + content', () => {
        const element = findStrategy('#some-class=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//*[contains(@id, "some-class") and normalize-space() = "some random text with "§$%&/()div=or others"]')
    })

    it('should find an element by tag name + id + similar content', () => {
        const element = findStrategy('div#some-id*=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//div[contains(@id, "some-id") and contains(., "some random text with "§$%&/()div=or others")]')
    })

    it('should find an element by id + similar content', () => {
        const element = findStrategy('#some-id*=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//*[contains(@id, "some-id") and contains(., "some random text with "§$%&/()div=or others")]')
    })

    it('should find an element by id + similar content see #1494', () => {
        const element = findStrategy('#What-is-WebdriverIO*=What')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//*[contains(@id, "What-is-WebdriverIO") and contains(., "What")]')
    })

    it('should find an element by tag name + attribute + content', () => {
        const element = findStrategy('div[some-attribute="some-value"]=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//div[contains(@some-attribute, "some-value") and normalize-space() = "some random text with "§$%&/()div=or others"]')
    })

    it('should find an element by attribute + content', () => {
        const element = findStrategy('[some-attribute="some-value"]=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//*[contains(@some-attribute, "some-value") and normalize-space() = "some random text with "§$%&/()div=or others"]')
    })

    it('should find an element by attribute existence + content', () => {
        const element = findStrategy('[some-attribute]=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//*[@some-attribute and normalize-space() = "some random text with "§$%&/()div=or others"]')
    })

    it('should find an element by tag name + attribute + similar content', () => {
        const element = findStrategy('div[some-attribute="some-value"]*=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//div[contains(@some-attribute, "some-value") and contains(., "some random text with "§$%&/()div=or others")]')
    })

    it('should find an element by attribute + similar content', () => {
        const element = findStrategy('[some-attribute="some-value"]*=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//*[contains(@some-attribute, "some-value") and contains(., "some random text with "§$%&/()div=or others")]')
    })

    it('should find an custom element by tag name + content', () => {
        const element = findStrategy('custom-element-with-multiple-dashes=some random text with "§$%&/()div=or others')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('.//custom-element-with-multiple-dashes[normalize-space() = "some random text with "§$%&/()div=or others"]')
    })

    it('should allow to go up and down the DOM tree with xpath', () => {
        let element = findStrategy('..')
        expect(element.using).toBe('xpath')
        element = findStrategy('.')
        expect(element.using).toBe('xpath')
        element = findStrategy('../')
        expect(element.using).toBe('xpath')
    })

    it('should find an element by ui automator strategy (android only)', () => {
        const element = findStrategy('android=foo')
        expect(element.using).toBe('-android uiautomator')
        expect(element.value).toBe('foo')
    })

    it('should find an element by datamatcher strategy (android only)', () => {
        const selector = { 'name': 'startsWith', 'args': 'aFakeString' }
        const element = findStrategy(selector)
        expect(element.using).toBe('-android datamatcher')
        expect(element.value).toBe('{"name":"startsWith","args":"aFakeString"}')
    })

    it('should find an element by viewmatcher strategy (android only)', () => {
        const selector = { 'name': 'startsWith', 'args': 'aFakeString', 'class': 'androidx.test.espresso.matcher.ViewMatchers' }
        const element = findStrategy(selector)
        expect(element.using).toBe('-android viewmatcher')
        expect(element.value).toBe('{"name":"startsWith","args":"aFakeString","class":"androidx.test.espresso.matcher.ViewMatchers"}')
    })

    it('should find an element by ui automation strategy (ios only)', () => {
        const element = findStrategy('ios=foo')
        expect(element.using).toBe('-ios uiautomation')
        expect(element.value).toBe('foo')
    })

    it('should find an element by predicate strategy (ios only)', () => {
        const element = findStrategy('-ios predicate string:type == \'XCUIElementTypeSwitch\' && name CONTAINS \'Allow\'')
        expect(element.using).toBe('-ios predicate string')
        expect(element.value).toBe('type == \'XCUIElementTypeSwitch\' && name CONTAINS \'Allow\'')
    })

    it('should find an element by class chain strategy (ios only)', () => {
        const element = findStrategy('-ios class chain:**/XCUIElementTypeCell[`name BEGINSWITH "D"`]/**/XCUIElementTypeButton')
        expect(element.using).toBe('-ios class chain')
        expect(element.value).toBe('**/XCUIElementTypeCell[`name BEGINSWITH "D"`]/**/XCUIElementTypeButton')
    })

    it('should find an element by accessibility id', () => {
        const element = findStrategy('~foo')
        expect(element.using).toBe('accessibility id')
        expect(element.value).toBe('foo')
    })

    it('should find an element by css selector with id and attribute', () => {
        const element = findStrategy('#purplebox[data-foundBy]')
        expect(element.using).toBe('css selector')
    })

    it('should find an element by css selector with id and immediately preceded operator', () => {
        const element = findStrategy('#purplebox+div')
        expect(element.using).toBe('css selector')
    })

    it('should find an element by css selector with id and preceded operator', () => {
        const element = findStrategy('#purplebox~div')
        expect(element.using).toBe('css selector')
    })

    it('should find an element by css selector with id and pseudo class', () => {
        const element = findStrategy('#purplebox:before')
        expect(element.using).toBe('css selector')
    })

    it('should find an element using "css selector" strategy if isMobile is false and w3c is used', () => {
        const element = findStrategy('[name="searchinput"]', true, false)
        expect(element.using).toBe('css selector')
        expect(element.value).toBe('[name="searchinput"]')
    })

    it('should find an element by android accessibility id', () => {
        const selector = 'new UiSelector().text("Cancel")).className("android.widget.Button")'
        const element = findStrategy('android=' + selector)
        expect(element.using).toBe('-android uiautomator')
        expect(element.value).toBe(selector)
    })

    it('should find an element by ios accessibility id', () => {
        const selector = 'UIATarget.localTarget().frontMostApp().mainWindow().buttons()[0]'
        const element = findStrategy('ios=' + selector)
        expect(element.using).toBe('-ios uiautomation')
        expect(element.value).toBe(selector)
    })

    it('should find an mobile ios element class name', () => {
        const element = findStrategy('UIATextField')
        expect(element.using).toBe('class name')
    })

    it('should find an mobile android element class name', () => {
        const element = findStrategy('android.widget.EditText')
        expect(element.using).toBe('class name')
    })

    it('should find an mobile youi.tv element class name', () => {
        const element = findStrategy('CYIPushButtonView')
        expect(element.using).toBe('class name')
    })

    it('allows to specify selector strategy directly', () => {
        let element = findStrategy('id:foobar id')
        expect(element.using).toBe('id')
        expect(element.value).toBe('foobar id')
        element = findStrategy('css selector:foobar css selector')
        expect(element.using).toBe('css selector')
        expect(element.value).toBe('foobar css selector')
        element = findStrategy('xpath:foobar xpath')
        expect(element.using).toBe('xpath')
        expect(element.value).toBe('foobar xpath')
        element = findStrategy('link text:foobar link text')
        expect(element.using).toBe('link text')
        expect(element.value).toBe('foobar link text')
        element = findStrategy('partial link text:foobar partial link text')
        expect(element.using).toBe('partial link text')
        expect(element.value).toBe('foobar partial link text')
        element = findStrategy('name:foobar name')
        expect(element.using).toBe('name')
        expect(element.value).toBe('foobar name')
        element = findStrategy('tag name:foobar tag name')
        expect(element.using).toBe('tag name')
        expect(element.value).toBe('foobar tag name')
        element = findStrategy('class name:foobar class name')
        expect(element.using).toBe('class name')
        expect(element.value).toBe('foobar class name')
        element = findStrategy('-android uiautomator:foobar -android uiautomator')
        expect(element.using).toBe('-android uiautomator')
        expect(element.value).toBe('foobar -android uiautomator')
        element = findStrategy('-android datamatcher:foobar -android datamatcher')
        expect(element.using).toBe('-android datamatcher')
        expect(element.value).toBe('foobar -android datamatcher')
        element = findStrategy('-android viewmatcher:foobar -android viewmatcher')
        expect(element.using).toBe('-android viewmatcher')
        expect(element.value).toBe('foobar -android viewmatcher')
        element = findStrategy('-android viewtag:foobar -android viewtag')
        expect(element.using).toBe('-android viewtag')
        expect(element.value).toBe('foobar -android viewtag')
        element = findStrategy('-ios uiautomation:foobar -ios uiautomation')
        expect(element.using).toBe('-ios uiautomation')
        expect(element.value).toBe('foobar -ios uiautomation')
        element = findStrategy('accessibility id:foobar accessibility id')
        expect(element.using).toBe('accessibility id')
        expect(element.value).toBe('foobar accessibility id')

        element = findStrategy('.ui-cloud__sec__develop-content__app-grid:nth-child(1)')
        expect(element.using).toBe('css selector')
        expect(element.value).toBe('.ui-cloud__sec__develop-content__app-grid:nth-child(1)')
    })

    it('should not allow unsupported selector strategies if w3c is used', () => {
        expect(() => findStrategy('accessibility id:foobar accessibility id', true)).toThrow()
        expect(() => findStrategy('android=foo', true)).toThrow()
    })

    it('should allow mobile selector strategies if isMobile is used', () => {
        let element = findStrategy('android=foo', undefined, true)
        expect(element.using).toBe('-android uiautomator')
        expect(element.value).toBe('foo')

        element = findStrategy('ios=foo', undefined, true)
        expect(element.using).toBe('-ios uiautomation')
        expect(element.value).toBe('foo')
    })

    it('should allow mobile selector strategies if isMobile is used even when w3c is used', () => {
        let element = findStrategy('android=foo', true, true)
        expect(element.using).toBe('-android uiautomator')
        expect(element.value).toBe('foo')

        element = findStrategy('ios=foo', true, true)
        expect(element.using).toBe('-ios uiautomation')
        expect(element.value).toBe('foo')
    })

    it('should find an mobile element using image string', () => {
        fs.readFileSync = jest.fn(() => 'random string') as any
        let element = findStrategy('/test.jpg')
        expect(element.using).toBe('-image')
        expect(element.value).toBe('random string')

        element = findStrategy('path/test.png')
        expect(element.using).toBe('-image')
        expect(element.value).toBe('random string')

        element = findStrategy('path/test.PNG')
        expect(element.using).toBe('-image')
        expect(element.value).toBe('random string')

        element = findStrategy('//path//test.svg')
        expect(element.using).toBe('-image')
        expect(element.value).toBe('random string')

        element = findStrategy('//xpath[@img="/test.png"]')
        expect(element.using).toBe('xpath')
        expect(element.value).not.toBe('random string')
    })
})
