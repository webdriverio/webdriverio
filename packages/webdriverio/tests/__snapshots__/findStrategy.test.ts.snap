// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`selector strategies helper > should be able to fetch element by aria label 1`] = `".//*[@aria-labelledby=(//*[normalize-space(text()) = "foobar"]/@id)] | .//*[@aria-describedby=(//*[normalize-space(text()) = "foobar"]/@id)] | .//*[@aria-label = "foobar"] | .//input[@id = (//label[normalize-space() = "foobar"]/@for)] | .//textarea[@id = (//label[normalize-space() = "foobar"]/@for)] | .//input[ancestor::label[normalize-space(text()) = "foobar"]] | .//textarea[ancestor::label[normalize-space(text()) = "foobar"]] | .//input[@placeholder="foobar"] | .//textarea[@placeholder="foobar"] | .//input[@aria-placeholder="foobar"] | .//textarea[@aria-placeholder="foobar"] | .//*[not(self::label)][@title="foobar"] | .//img[@alt="foobar"] | .//*[not(self::label)][normalize-space(text()) = "foobar"]"`;
